File: src/core/apply.ts
import type { GameState, Move } from './types';
import { opponent } from './rules';

/** 합법 수라고 가정하고 적용한다. 검증은 legalMoves 쪽 책임. */
export function applyMove(state: GameState, move: Move): GameState {
  const board = state.board.map((row) => row.slice());
  const guardsInHand = { ...state.guardsInHand };

  if (move.kind === 'PLACE') {
    board[move.to.r][move.to.c] = { player: state.turn, type: 'GUARD' };
    guardsInHand[state.turn] -= 1;
  } else {
    const piece = board[move.from.r][move.from.c]!;
    board[move.from.r][move.from.c] = null;
    board[move.to.r][move.to.c] = piece; // 상대 호위가 있으면 대체 잡기
  }

  return {
    board,
    turn: opponent(state.turn),
    guardsInHand,
    history: [...state.history, move],
  };
}

File: apps/ios/Sources/MongjinCore/Apply.swift
import Foundation

/// 합법 수라고 가정하고 적용한다. 검증은 `legalMoves` 쪽 책임.
public func applyMove(_ state: GameState, _ move: Move) -> GameState {
    var board = state.board
    var guardsInHand = state.guardsInHand

    switch move {
    case .place(let to):
        board[to.r][to.c] = Piece(player: state.turn, type: .escort)
        guardsInHand[state.turn, default: 0] -= 1
    case .move(let from, let to):
        let piece = board[from.r][from.c]
        board[from.r][from.c] = nil
        board[to.r][to.c] = piece
    }

    var next = GameState(
        board: board,
        turn: state.turn.opponent,
        guardsInHand: guardsInHand,
        history: state.history + [move],
        positionCounts: state.positionCounts
    )
    let key = positionKey(next)
    next.positionCounts[key, default: 0] += 1
    return next
}

/// 탐색용: history·positionCounts를 비운 가벼운 자식 국면
public func childState(_ state: GameState, _ move: Move) -> GameState {
    var board = state.board
    var guardsInHand = state.guardsInHand
    switch move {
    case .place(let to):
        board[to.r][to.c] = Piece(player: state.turn, type: .escort)
        guardsInHand[state.turn, default: 0] -= 1
    case .move(let from, let to):
        let piece = board[from.r][from.c]
        board[from.r][from.c] = nil
        board[to.r][to.c] = piece
    }
    return GameState(
        board: board,
        turn: state.turn.opponent,
        guardsInHand: guardsInHand,
        history: [],
        positionCounts: [:]
    )
}
